### ISSUE TITLE
Question : difference in cpu resize and gpu resize

### ISSUE THREAD

**Issue by emmawoo6** [NVIDIA/DALI#1602]
Hello, I've just started coding with dali, and I'm interested in the resize function. So I just compared different inputs and the results. My codes is listed below. I'm not using CropMirrorNormalize after resize.

It turns out that the binary file generated by "cpu linear" and "gpu linear" are slitghlty different (+/-1 bit). And "cpu linear" is different from the one generated by OpenCV resize (+/-1 bit , but the ratio is higher).

Can you expain the reason why there is difference between them? Also, I've read some issues on the drop of accuracy caused by the resize algorithm. Do you think it's important to keep the results 100% identical to OpenCV results? Or it's okay to have such difference - if developers utilize the same algorithm during training and validation, the accuracy will not be affected.
1.  cpu, linear
```
class SimplePipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id):
        super(SimplePipeline, self).__init__(batch_size, num_threads, device_id, seed = 12)
        self.input = ops.FileReader(file_root = image_dir)
        # instead of path to file directory file with pairs image_name image_label_value can be provided
        # self.input = ops.FileReader(file_root = image_dir, file_list = image_dir + '/file_list.txt')
        self.decode = ops.ImageDecoder(device = 'cpu', output_type = types.RGB)
        self.resize = ops.Resize(device="cpu", resize_x = 416, resize_y = 416,interp_type=types.INTERP_LINEAR)
        self.cmnp = ops.CropMirrorNormalize(device="gpu",
						crop=(416, 416),
						output_dtype=types.FLOAT,
						output_layout=types.NHWC,
						image_type=types.RGB,
						# mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
						# std=[0.229 * 255, 0.224 * 255, 0.225 * 255]
						)       
        self.coin = ops.CoinFlip(probability = 0.5)

    def define_graph(self):
        rng = self.coin()
        jpegs, labels = self.input()
        images = self.decode(jpegs)
        images = self.resize(images)
        # output = self.cmnp(images)
        return (images, labels)

def main(argv):
    pipe = SimplePipeline(batch_size, 1, 0)
    pipe.build()
    
    pipe_out = pipe.run()
    print(pipe_out)

    images, labels = pipe_out
    print("Images is_dense_tensor: " + str(images.is_dense_tensor()))
    print("Labels is_dense_tensor: " + str(labels.is_dense_tensor()))
    print(np.array(images.at(0))[0])
    print(np.shape(np.array(images.at(0))))
    img0 = np.array(images.at(0))
    img0.tofile("img0_linear_cpu.bin")
```

2. gpu, linear
```
class SimplePipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id):
        super(SimplePipeline, self).__init__(batch_size, num_threads, device_id, seed = 12)
        self.input = ops.FileReader(file_root = image_dir)
        # instead of path to file directory file with pairs image_name image_label_value can be provided
        # self.input = ops.FileReader(file_root = image_dir, file_list = image_dir + '/file_list.txt')
        self.decode = ops.ImageDecoder(device = 'cpu', output_type = types.RGB)
        self.resize = ops.Resize(device="gpu", resize_x = 416, resize_y = 416,interp_type=types.INTERP_LINEAR)
        self.cmnp = ops.CropMirrorNormalize(device="gpu",
						crop=(416, 416),
						output_dtype=types.FLOAT,
						output_layout=types.NHWC,
						image_type=types.RGB,
						# mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
						# std=[0.229 * 255, 0.224 * 255, 0.225 * 255]
						)       
        self.coin = ops.CoinFlip(probability = 0.5)

    def define_graph(self):
        rng = self.coin()
        jpegs, labels = self.input()
        images = self.decode(jpegs)
        images = self.resize(images.gpu())
        # output = self.cmnp(images)
        return (images, labels)

def main(argv):
    pipe = SimplePipeline(batch_size, 1, 0)
    pipe.build()
    
    pipe_out = pipe.run()
    print(pipe_out)

    images, labels = pipe_out
    print("Images is_dense_tensor: " + str(images.is_dense_tensor()))
    print("Labels is_dense_tensor: " + str(labels.is_dense_tensor()))
    print(np.array(images.as_cpu().at(0))[0])
    print(np.shape(np.array(images.as_cpu().at(0))))
    img0 = np.array(images.as_cpu().at(0))
    #print(np.array(images.at(0))[0])
    #print(np.shape(np.array(images.at(0))))
    #img0 = np.array(images.at(0))
    img0.tofile("img0_linear_cpu.bin")
```

**Comment by JanuszL** [comment]
@emmawoo6 that is an excellent question.
The reason behind the CPU and GPU implementation difference is the floating-point arithmetic. It is very hard to make GPU and CPU to apply all the operations in the same order to get the bit-exact results (due to different optimizations, compiler or even implementation strategy selected on what works best for CPU and what for GPU). So it is expected.
I think the OpenCV is not the best tool, to be honest. One of the reasons it is not flip invariant - it means that flip(resize(flip)) doesn't return the same result as just resize. Also, OpenCV resize is using fixed point interpolation coefficients which may lead to bigger rounding errors.

Yes, if you are using the same algorithm for validation and training it should not make any difference. However, if you are using presized images, that may have some aliasing artifacts triangular interpolation may not return the best results comparing to linear one (another story is when you have raw training data and presized validation one, then it is very hard to match).

**Comment by emmawoo6** [comment]
> @emmawoo6 that is an excellent question.
> The reason behind the CPU and GPU implementation difference is the floating-point arithmetic. It is very hard to make GPU and CPU to apply all the operations in the same order to get the bit-exact results (due to different optimizations, compiler or even implementation strategy selected on what works best for CPU and what for GPU). So it is expected.
> I think the OpenCV is not the best tool, to be honest. One of the reasons it is not flip invariant - it means that flip(resize(flip)) doesn't return the same result as just resize. Also, OpenCV resize is using fixed point interpolation coefficients which may lead to bigger rounding errors.
> 
> Yes, if you are using the same algorithm for validation and training it should not make any difference. However, if you are using presized images, that may have some aliasing artifacts triangular interpolation may not return the best results comparing to linear one (another story is when you have raw training data and presized validation one, then it is very hard to match).

Thank you for your reply, it's very inspiring ~ 
In that case, can dali resize algorithm overcome these shortcomings? Maybe I shall read the code to learn more about it.

**Comment by JanuszL** [comment]
> Thank you for your reply, it's very inspiring ~
> In that case, can dali resize algorithm overcome these shortcomings? Maybe I shall read the code to learn more about it.

That was one of the ideas when we developed it, and we recommend to use DALI as you have bit-exact results in different FWs (TensorFlow, MXNet, PyTorch and PaddlePaddle). The code itself is not that straightforward due to generalization and optimization. But if you want to know the guts feel free to do so.

**Comment by emmawoo6** [comment]
> > Thank you for your reply, it's very inspiring ~
> > In that case, can dali resize algorithm overcome these shortcomings? Maybe I shall read the code to learn more about it.
> 
> That was one of the ideas when we developed it, and we recommend to use DALI as you have bit-exact results in different FWs (TensorFlow, MXNet, PyTorch and PaddlePaddle). The code itself is not that straightforward due to generalization and optimization. But if you want to know the guts feel free to do so.

Thank you~ I'll try it.
